jitstats.lua

```diff
@@ -0,0 +1,462 @@
+    old = old[key]
+    new = new[key]
+    result = result[key]
+
+    for k, v in pairs(new) do
+        local change = v - (old[k] or 0)
+        if change ~= 0 then
+            result[k] = change
+        end
+    end
+    return result
+end
+
+-- Get the diff between two jit stat snapshots
+function jitstats.diffsnapshots(snap1, snap2)
+    local result = {
+        abort_counts = {},
+        nyi_bc = {},
+        linktypes = {},
+    }
+
+    for k, v in pairs(snap2) do   
+        if type(v) == "number" then
+            result[k] = v - (snap1[k] or 0)
+        end
+    end
+
+    diffstatbl(snap1, snap2, result, "abort_counts")
+    diffstatbl(snap1, snap2, result, "nyi_bc")
+    diffstatbl(snap1, snap2, result, "linktypes")
+
+    return result
+end
+
+local last_starts, last_exits
+
+function jitstats.stats_changed()
+
+    if last_starts == stats.starts and last_exits == stats.exits then
+        return false
+    end
+
+    last_starts = stats.starts
+    last_exits = stats.exits
+    return true
+end
+
+function jitstats.setdebugprint(enable)
+    jitstats.debug = enable
+    debugprint = (enable and printf) or function() end
+end
+
+local function topercent(total, subval)
+    return math.floor(subval / total * 100)
+end
+
+function jitstats.print_simple(msg)
+    local running = jitstats.stop()
+    printf("%sStarted %d, Aborted %d, Exits %d", msg or "", stats.starts, stats.aborts, stats.exits)
+    if running then
+        jitstats.start()
+    end
+end
+
+local function table_keys(t)
+    local result = {}
+    for k,_ in pairs(t) do
+        table.insert(result, k)
+    end
+    return result
+end
+
+local function default_sorter(t, k1, k2)
+    return t[k1] > t[k2] 
+end
+
+local function table_sortkeys(t, sorter)
+    local keys = table_keys(t) 
+    sorter = sorter or default_sorter
+    table.sort(keys, function(k1, k2) return sorter(t, k1, k2) end)
+    return keys
+end
+
+local function statstbl_concat(tbl)
+    local strings = {}
+    local count = 0
+    for k, v in pairs(tbl) do
+        count = count + 1
+        strings[count] = k.." = "..v
+    end
+    
+    return table.concat(strings, ", ")
+end
+
+function jitstats.print(statstbl, msg)
+    local running = jitstats.stop()
+
+    if not statstbl then
+        statstbl = stats
+        msg = "Trace Stats:"
+    elseif type(statstbl) == "string" then
+        assert(msg == nil)
+        msg = statstbl
+        statstbl = stats
+    else
+        assert(type(statstbl) == "table", type(statstbl))
+        msg = "Trace Stats:"
+    end
+
+    printf(msg)
+    printf("  Started %d, side %d(%g%%), stitch %d", statstbl.starts, statstbl.side_starts, topercent(statstbl.starts, statstbl.side_starts), statstbl.stitch_starts)
+    printf("  Completed Link types: %s", statstbl_concat(statstbl.linktypes))
+    printf("  Aborted %d(%g%%), side %d, stitch %d", statstbl.aborts, topercent(statstbl.starts, statstbl.aborts), statstbl.side_aborts, statstbl.stitch_aborts)
```

Comment:
this is hard to maintain.  what about a simplistic template?  even just like this (untested!):
```
print(([[
   Started {starts}, side {side_starts}({pct_sides}%), ...
   ...
]]):gsub('{([_%w]+)}', copy_table(statstbl, {
    pct_sides = topercent(statstbl.starts, statstbl.side_starts),
    ...
})
```

Reply:
I was thinking before of making topercent actually returning a string format of %d(%g%%) to simplify things. I have made my own poor man’s moustache before https://github.com/fsfod/LuaJIT/blob/gcperflog/src/perflog_apigen.lua#L43

Reply:
haven't we all?  (tip: use a capture group and you can lose the `key:sub(3,-3)`, and the whole function is just a table)
but the thing with these "lots of printf()s" blocks is that when you add data fields or change the phrasing, you start to loose count of which %d is which value.  that's when a `:gsub({{...}}, values)` comes handy

Reply:
fixed with a template system for the main unconditional block of text